calculator/src/main/java/com/market/banica/calculator/service/TransactionServiceImpl.java

```diff
@@ -0,0 +1,97 @@
+import com.market.banica.calculator.service.contract.CalculatorService;
+import com.market.banica.calculator.service.contract.TransactionService;
+import com.market.banica.calculator.service.grpc.AuroraClientSideService;
+import com.market.banica.common.exception.ProductNotAvailableException;
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class TransactionServiceImpl implements TransactionService {
+    private static final String PRODUCT_NOT_AVAILABLE_MESSAGE = "%s with quantity %d is currently not available on the %s market";
+
+    private final CalculatorService calculatorService;
+    private final AuroraClientSideService auroraClientSideService;
+
+    @Override
+    public List<ProductDto> buyProduct(String clientId, String itemName, long quantity) throws ProductNotAvailableException {
+        List<ProductDto> purchaseProducts = this.calculatorService.getProduct(clientId, itemName, quantity);
+
+        List<ProductDto> notCompoundProducts = getNotCompoundProducts(purchaseProducts);
+
+        List<ItemDto> pendingItems = new ArrayList<>();
+
+        boolean areAvailable = true;
+        String productAvailabilityResponseErrorMessage = "";
+
+        for (ProductDto purchaseProduct : notCompoundProducts) {
+            if (!areAvailable) {
+                break;
+            }
+            String productName = purchaseProduct.getItemName();
+
+            for (ProductSpecification productSpecification : purchaseProduct.getProductSpecifications()) {
+
+                BigDecimal productPrice = productSpecification.getPrice();
+                Long productQuantity = productSpecification.getQuantity();
+                Origin productOrigin = Origin.valueOf(productSpecification.getLocation().toUpperCase(Locale.ROOT));
+
+                AvailabilityResponse availabilityResponse = this.auroraClientSideService.checkAvailability(productName, productPrice.doubleValue(), productQuantity, productOrigin);
+
+                if (!availabilityResponse.getIsAvailable()) {
+                    areAvailable = false;
+                    productAvailabilityResponseErrorMessage = String.format(PRODUCT_NOT_AVAILABLE_MESSAGE, availabilityResponse.getItemName(), availabilityResponse.getItemQuantity(), availabilityResponse.getMarketName());
+                    break;
+                }
+                pendingItems.add(new ItemDto(productName, productPrice, productOrigin.toString(), productQuantity));
+            }
+        }
+
+        if (!areAvailable) {
+            returnPendingProducts(pendingItems);
+
+            throw new ProductNotAvailableException(productAvailabilityResponseErrorMessage);
```

Comment:
since you are throwing an exception, you can move directly returnPendingProducts and throwing the exception on row 56 - it will exit the for cycles when throwing the exception is reached. This way you can get rid of areAvailable flag too